src/pod_util.ts

```diff
@@ -0,0 +1,240 @@
+  Object.values(getStateConstants()).concat(regionsList).sort()
+);
+
+const getValidStateAndRegionsSet = memoize(
+  () => new Set(listStateAndRegions())
+);
+
+const isValidStateOrRegionName = (stateOrRegionName: string) =>
+  getValidStateAndRegionsSet().has(stateOrRegionName);
+
+function getRedisKeyWithStateOrRegionName(
+  stateOrRegionName: string,
+  channelType: CHANNEL_TYPE,
+  entryPointType: ENTRYPOINT_TYPE
+) {
+  const nameNoSpaces = stateOrRegionName.replace(/\s+/g, '');
+  const entrypointString =
+    entryPointType === ENTRYPOINT_TYPE.PULL ? 'Pull' : 'Push';
+  switch (channelType) {
+    case CHANNEL_TYPE.NORMAL:
+      return `openPods${entrypointString}${nameNoSpaces}`;
+    case CHANNEL_TYPE.DEMO:
+      return `openPods${entrypointString}Demo${nameNoSpaces}`;
+  }
+  throw new Error(`Unexpected channelType: ${channelType}`);
+}
+
+export function getChannelNamePrefixForStateOrRegionName(
+  stateOrRegionName: string,
+  channelType: CHANNEL_TYPE
+): string {
+  const nameWithHyphens = stateOrRegionName.toLowerCase().replace(/\s+/g, '-');
+  switch (channelType) {
+    case CHANNEL_TYPE.NORMAL:
+      return `${nameWithHyphens}-`;
+    case CHANNEL_TYPE.DEMO:
+      return `demo-${nameWithHyphens}-`;
+  }
+  throw new Error(`Unexpected channelType: ${channelType}`);
+}
+
+async function getPodChannelStateForEntrypoint(
+  redisClient: PromisifiedRedisClient,
+  channelNamesAndIds: Record<string, string>,
+  filters: PodFilters,
+  entrypoint: ENTRYPOINT_TYPE
+) {
+  // Map from channel names to a state
+  const ret: Record<string, ChannelInfo> = {};
+
+  const openChannels = await redisClient.lrangeAsync(
+    getRedisKeyWithStateOrRegionName(
+      filters.stateOrRegionName,
+      filters.channelType,
+      entrypoint
+    ),
+    0,
+    -1
+  );
+
+  // Helper to make default channel info object
+  const makeChannelInfo = (channelName: string) => ({
+    id: channelNamesAndIds[channelName],
+    channelName,
+    entrypoint,
+    weight: 0,
+  });
+
+  // Count all the open channels
+  openChannels.forEach((channelName) => {
+    ret[channelName] = ret[channelName] || makeChannelInfo(channelName);
+    ret[channelName].weight += 1;
+  });
+
+  // Assign weight 0 to any remaining channels.
+  Object.keys(channelNamesAndIds).forEach((channelName: string) => {
+    if (
+      channelName.startsWith(
+        getChannelNamePrefixForStateOrRegionName(
+          filters.stateOrRegionName,
+          filters.channelType
+        )
+      )
+    )
+      ret[channelName] = ret[channelName] || makeChannelInfo(channelName);
+  });
+
+  return sortBy(ret, 'channelName');
+}
+
+export async function getPodChannelState(
+  redisClient: PromisifiedRedisClient,
+  filters: PodFilters
+): Promise<{ pull: ChannelInfo[]; push: ChannelInfo[] }> {
+  // Get all channels from Redis cache or Slack API
+  const channelNamesAndIds =
+    (await SlackApiUtil.getSlackChannelNamesAndIds(redisClient)) || {};
```

Comment:
One thing to note is that the cache used in `getSlackChannelNamesAndIds` is currently only refreshed when a voter is routed to a channel that doesn't exist in the cache, either automatically from the lobby or manually via ROUTE_VOTER. The former creates a catch 22 (can't open the channel because it's not in the cache because a voter hasn't been routed there because the channel isn't open because it can't be opened because it's not on the modal because it's not in the cache etc...).

https://github.com/tomerovadia/VoterHelpLine/blob/96b3f90c3f01a53742b79705ebdf015325e4a819/src/router.ts#L512

Maybe you can modify `getSlackChannelNamesAndIds` to take a parameter that indicates whether the cache should be forced to update?

Reply:
I was thinking we have two ways to go about this:

1. Add a command to manually invalidate the cache as needed.
2. Listen for Slack events to invalidate or update the cache whenever a new channel is created.

I can add that in a follow up PR.